Add check for possible CStr literals in pre-2021

chfogelman opened the pull request:
Fixes [#118654](https://github.com/rust-lang/rust/issues/118654)

Adds information to errors caused by possible CStr literals in pre-2021.

The lexer separates `c"str"` into two tokens if the edition is less than 2021, which later causes an error when parsing. This error now has a more helpful message that directs them to information about editions. However, the user might also have written `c "str"` in a later edition, so to not confuse people who _are_ using a recent edition, I also added a note about whitespace.

We could probably figure out exactly which scenario has been encountered by examining spans and editions, but I figured it would be better not to overcomplicate the creation of the error too much.

This is my first code PR and I tried to follow existing conventions as much as possible, but I probably missed something, so let me know!

chfogelman committed:
Improve compiler error for c-strings in pre-2021

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,17 @@ impl<'a> Parser<'a> {
         }
 
+        // extra info for `c"str"` before 2021 edition or `c "str"` in all editions
+        if self.prev_token.is_ident_named(Symbol::intern("c"))
```
```suggestion
        if self.prev_token.is_ident_named(sym::c)
```

fmease replied:
You probably also want to check for `cr` (as in `cr""`, raw c-string literals). If I'm not mistaken “`sym::c && StrRaw`” only accounts for `c r""` (notice the space).

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,17 @@ impl<'a> Parser<'a> {
+        // extra info for `c"str"` before 2021 edition or `c "str"` in all editions
+        if self.prev_token.is_ident_named(Symbol::intern("c"))
+            && let TokenKind::Literal(token::Lit { kind: token::Str | token::StrRaw(..), .. }) =
+                &self.token.kind
```
You could check if the `c` immediately precedes the string literal I guess via `self.prev_token.span.hi() == self.token.span.lo()` to rule out the false positives like `c ""` (notice the space).

fmease replied:
While you noted

> We could probably figure out exactly which scenario has been encountered by examining spans and editions, but I figured it would be better not to overcomplicate the creation of the error too much.

it should be as simple as `self.prev_token.span.eq_ctxt(self.token.span) && self.token.span.edition < Edition::Edition2021`.

The `eq_ctxt` check prevents the notes about cstr literals from misfiring on:

```rs
// edition: <2021
macro_rules! m { ($x:ident) => { $x"" } }
fn main() { m!(c) }
```

Your current implementation fires here.

chfogelman replied:
I'm fixing this up, but using this macro as an example, `self.prev_token.span.eq_ctxt(self.token.span) == true`. The previous token's span appears to be that of the `$x` token in the macro expansion body, rather than the `c` argument to the macro, so they have the same context. Is this correct, or am I missing something?

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,17 @@ impl<'a> Parser<'a> {
+                &self.token.kind
+        {
+            err.note("you may be trying to declare a CStr literal");
+            err.note("`CStr` literals require edition 2021");
```
Backticks in diagnostics are for quoting code and user-written text, not for names.
```suggestion
            err.note("`c-string literals require at least edition 2021");
```

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,17 @@ impl<'a> Parser<'a> {
+            && let TokenKind::Literal(token::Lit { kind: token::Str | token::StrRaw(..), .. }) =
+                &self.token.kind
+        {
+            err.note("you may be trying to declare a CStr literal");
```
You could think about making this diagnostic translatable. https://rustc-dev-guide.rust-lang.org/diagnostics/translation.html#writing-translatable-diagnostics

chfogelman commented:
I've gone ahead and removed the ambiguity with whitespace, leaving that to just be a standard "expected one of" errors for consistency with `b"str"` and others. I added some additional checks to ensure that what we have is _really_ intended to be a cstr literal, and also to pick up raw cstr literals and eliminate false positives in macro expansion. And I changed the error message itself to remove the operator soup and just say that the string token itself was unexpected.

This could still misfire on the characters `cr#` if they are not the start of a c-string, but I'm not sure that's worth trying to get perfect. At this point, the lexer has already turned this into the tokens [`cr`, `#`,...], so this would essentially require us to re-lex the rest of the statement to be 100% sure we're looking at something that _would be_ a cstr literal in later editions.

fmease commented:
r? fmease

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
             }
         }
 
+        // Extra info for `c"str"` before 2021 edition or `c "str"` in all editions. The heuristic
```
```suggestion
        // Try to detect an attempt by the user to write a c-string literal before the 2021 edition. The heuristic
```

or something like that (it gives a better summary imo). Note that your comment still mentions `c "str"` (with a space) which is no longer accurate.

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
+        // edition where c-string literals are not allowed. There is the slight possibility of a
+        // false positive for a `cr#` that wasn't intended to start a c-string literal, but the
+        // lexer was greedy and didn't preserve whether the `r#` on its own would have started a
+        // valid raw string literal.
```
This sentence is fine but you could also think about expanding upon it with something along the lines of “and we don't want to perform unbounded lookahead here to check if we have a sequence of hashes followed by a string literal”.

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
+            && self.prev_token.span.hi() == self.token.span.lo()
+            && !self.token.span.at_least_rust_2021()
+        {
+            err.cancel();
```
Actually, since this subdiagnostic can have false positives, it might be wiser to keep the “token soup” error, not sure. When I suggested canceling the error, I assumed the diagnostic would be 100% accurate and we could even replace the main message of the diagnostic similar to `error[E0670]: `async fn` is not permitted in Rust 2015`.

Your call though. If you'd like to keep the shorter message, I'd go with `unexpected token` (dropping the `found`) and with `unexpected token` over `found here` since the current phrasing isn't used in any other of rustc's error messages or only very rarely regarding the latter.

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
+            let mut err =
+                self.struct_span_err(self.token.span, format!("found unexpected token {descr}"));
+            err.span_label(self.token.span, "found here".to_owned());
+            err.note("you may be trying to declare a c-string literal");
```
```suggestion
            err.note("you might have meant to write a c-string literal");
```

Not a native speaker but the word *declare* doesn't feel right to me in this context.

fmease commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
+                self.struct_span_err(self.token.span, format!("found unexpected token {descr}"));
+            err.span_label(self.token.span, "found here".to_owned());
+            err.note("you may be trying to declare a c-string literal");
+            err.note("c-string literals require edition 2021 or later");
```
```suggestion
            err.note("c-string literals require the 2021 edition or later");
```

```suggestion
            err.note("c-string literals require Rust 2021 or later");
```

(minor) Preexisting diagnostics always seem to be using one of the suggested expressions.

fmease reviewed (approved):
Thanks, that's great! Sorry for taking so long. Your code looks good, I just have some nitpicks about the comments and the diagnostics.

Lastly, could you squash your commits in one? After that, it should be ready to go!

chfogelman commented:
Okay, went ahead and reverted to token soup because I agree it's more appropriate for this case; we really just want a hint here, not a whole new error (also saves me the trouble of being too creative with error messages). Also changed `declare`->`write`, and `edition`->`Rust`, but left the other phrasing as-is. And sqashed. Thanks for the reviews!

fmease commented:
Thanks! :)

@bors r+ rollup

bors commented:
:pushpin: Commit 2c9602587472a94afc595f0f5adc7a0b85e15201 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 2c9602587472a94afc595f0f5adc7a0b85e15201 -->
<!-- homu: {"type":"Approved","sha":"2c9602587472a94afc595f0f5adc7a0b85e15201","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
